Plugins: Refresh plugin-install.php instead of redirect on activation. (v2)

costdev opened the pull request:
This PR implements an alternative approach to #6717.

- Adds a `$context` parameter to `wp_get_plugin_action_button()` to conditionally modify the button's markup.
- When generating the `Activate` button, adds a `redirect_to` query arg with the URL-encoded referer as the value.
- Passes `'plugin-information-modal'` as the context when generating the button inside the modal.
- When the referer is `plugin-install.php`, redirects to the `redirect_to` query arg's value, if present, or to the referrer.
  - `redirect_to` covers modal-based activation.
  - referer covers card-based activation.
- Ensures the JS-generated `Activate` button contains the `redirect_to` query arg, only if in the modal.
- Ensures the user receives feedback both on `plugins.php` and `plugin-install.php`.

**Note: Remember to build the PR branch and perform a hard refresh on `plugins.php` or `plugin-install.php` before testing.**

Trac ticket: https://core.trac.wordpress.org/ticket/61330

costdev committed:
Introduce a `$context` parameter to `wp_get_plugin_action_button()`.

costdev committed:
Add a `redirect_to` parameter for `plugin-information-modal` context.

costdev committed:
Pass `plugin-information-modal` context when in a plugin's modal.

costdev committed:
Redirect to `redirect_to` or the referer, when coming from `plugin-install.php`.

costdev committed:
JS: Add the `redirect_to` query arg when generating `Activate` button.

mukeshpanchal27 commented on src/wp-admin/includes/plugin-install.php:
```diff
@@ -912,9 +912,10 @@ function install_plugin_information() {
  * }
  * @param bool         $compatible_php   The result of a PHP compatibility check.
  * @param bool         $compatible_wp    The result of a WP compatibility check.
+ * @param string       $context          Optional. The context the button will appear in. Default empty string.
```
Needs to add `@since` for new variable.

costdev replied:
Thanks @mukeshpanchal27! I held off on this for now as it's still in an evaluation stage. Still good to have that reminder if we forget as things move forward! I'll add a `@since` annotation just so we have something in place to update.

costdev replied:
Updated and force-pushed! 👍

costdev committed:
Provide user feedback for successful activation on `plugins.php`.

costdev committed:
Provide user feedback for successful activation on `plugin-install.php`.

costdev commented on src/wp-admin/plugin-install.php:
```diff
@@ -24,6 +24,17 @@
+			'dismissible'        => true,
+		)
+	);
+}
```
In case you're wondering why the `updated` class is added, this is what's used for [the notice](https://github.com/wordpress/wordpress-develop/blob/6.5/src/wp-admin/plugins.php#L723) on `plugins.php`, which uses [this array](https://github.com/wordpress/wordpress-develop/blob/6.5/src/wp-admin/plugins.php#L646). For consistency and to reduce the chances of unexpected differences, I used the same array data in `plugin-install.php`.

hellofromtonya commented on src/wp-admin/includes/plugin-install.php:
```diff
@@ -912,9 +913,10 @@ function install_plugin_information() {
  * }
  * @param bool         $compatible_php   The result of a PHP compatibility check.
  * @param bool         $compatible_wp    The result of a WP compatibility check.
+ * @param string       $context          Optional. The context the button will appear in. Default empty string.
```
What do you think about adding what values it accepts?

For example, `sanitize_bookmark_field()` is documented as:

```
 * @param string $context     How to filter the field value. Accepts 'raw', 'edit', 'db',
 *                            'display', 'attribute', or 'js'. Default 'display'.
 ```
 
I'm thinking this will help devs and contributors.

costdev replied:
Maybe, though at the moment we only have 1 entry so it'd be something like:

> Accepts 'plugin-information-modal'. Default empty string.

I'm not sure how useful that is at the moment, but that's likely just because of limited values currently, and a possible filter hook in 6.7 could change this.

I'm happy to add to the description. Does the above work for you, or do you an alternative suggestion?

hellofromtonya replied:
I like that description. It's very clear what that param accepts. Any other value is not (currently) used.

hellofromtonya replied:
It's useful in that devs / contributors will know that this param only accepts that value. In the future, more values will likely be added.

costdev replied:
Implemented in [30832d9](https://github.com/WordPress/wordpress-develop/pull/6732/commits/30832d9762e82e5d8725c08262e83a421188ce6f).

costdev committed:
Update `$context` parameter description.

costdev committed:
Move `plugin-install.php` notice to avoid "headers already sent" warning.

hellofromtonya commented on src/wp-admin/plugin-install.php:
```diff
@@ -24,6 +24,17 @@
 	exit;
 }
 
+if ( isset( $_GET['activate'] ) ) {
+	wp_admin_notice(
+		__( 'Plugin activated.' ),
+		array(
+			'id'                 => 'message',
+			'additional_classes' => array( 'updated' ),
+			'dismissible'        => true,
+		)
+	);
+}
```
A Warning for headers already sent happens with some plugins (such as bbPress, Custom Twitter Feeds, etc) and inconsistently with many others.

In testing with @costdev, found that this new admin notice is the likely suspect. Commenting this out and/or moving it to the end of the file (before the footer) resolves the warning.

costdev replied:
Relocated in [21e4899](https://github.com/WordPress/wordpress-develop/pull/6732/commits/21e4899565c9ddb1a493d26bb055f1eaa6c8b221).

costdev committed:
Adjust referer check to account for multisite.

#BlameJorbin
unprops @aaronjorbin
unjorb propsin

costdev committed:
Ensure that `load-plugins.php` hook fires. (read more)

Previously, this PR redirected to either the previous page, or the URL designated in the `redirect_to` query argument. This redirect occurred immediately after the plugin was activated.

Testing revealed that if a plugin hooked a redirect, or other functionality, on `load-plugins.php`, then this wasn't fired after plugin activation.

To ensure that the `load-plugins.php` hook, and anything else that's not specifically tied to the output of `plugins.php` markup, this change introduces a new `finish-activation` action.

Upon activation, a `redirect_to` query argument is added with the value either of the previous page, or of an existing `redirect_to` value.

If the plugin's activation was successful, an `activate` query argument is also added with the value `'true'`, to ensure feedback is provided to users in the form of an admin notice (already existing in Core).

As these redirects are performed using `wp_redirect()`, this happens during the same HTTP request. As there is no output prior to the redirects, `headers already sent` warnings should not occur.

Test, test, test. Many plugins, single site and multisite, and any other testing paths you can think of.

costdev committed:
Only `exit` on redirect. Otherwise, `break`.

costdev committed:
Avoid double-encoding URLs when adding query arguments.

`add_query_arg()` calls `urlencode_deep()` internally.

costdev commented on src/wp-admin/plugins.php:
```diff
@@ -87,10 +87,10 @@
 				if ( isset( $_GET['redirect_to'] ) ) {
 					// This will occur when a plugin is activated from within a modal.
-					$url = add_query_arg( 'redirect_to', urldecode( $_GET['redirect_to'] ), $url );
+					$url = add_query_arg( 'redirect_to', $_GET['redirect_to'], $url );
```
@peterwilsoncc Security review request. Should this be run through `wp_unslash()`?

costdev committed:
Run all `$_GET` values through `wp_unslash()`.

costdev committed:
Add some inline comments.

costdev committed:
Add explainer for the `activate` query argument.

costdev committed:
Do not unnecessarily `urlencode()` the referrer.

`add_query_arg()` calls `urlencode_deep()` internally.

costdev committed:
Only redirect if a previous redirect was not performed.

hellofromtonya commented on src/wp-admin/plugins.php:
```diff
@@ -83,12 +82,56 @@
+					$url = add_query_arg( 'activate', 'true', $url );
+				}
 
+				wp_redirect( $url );
```
Found an issue. Adding it here for awareness and discussion:

When using a search query, after activating a plugin (that does not have its own redirect), the user is redirected to Add Plugins UI without the search results.

Steps:
1. Go to Add Plugins UI.
2. In the Search plugins field, type `core rollback`.
3. Install > Activate

Expected results:
It should activate and then redirect back to the same page with the search results.

Before activation, here's the URL:
```
wp-admin/plugin-install.php?s=core%2520rollback&tab=search&type=term
```

After activation, here's the URL:
```
wp-admin/plugin-install.php?s=core+rollback
```

costdev replied:
Interesting that those query arguments are lost. 🤔

hellofromtonya replied:
Test report is here https://core.trac.wordpress.org/ticket/61330#comment:31

costdev replied:
Fixed in [ae9b540](https://github.com/WordPress/wordpress-develop/pull/6732/commits/ae9b540c7c2166e43a6a87e8bd357d13dbfa6948).

Commit message:
> Encode the referrer to preserve multiple query arguments.
> 
> Without URL-encoding, only the first `?arg=value` is retained, and any `&argX=valueX` is lost. Wrapping `wp_get_referer()` in `urlencode()` preserves these additional query arguments.

hellofromtonya replied:
Thanks @costdev. Retesting now...

hellofromtonya replied:
Yes, the fix works. Tested it in each tab and with search results. After activation of a plugin with no redirect in it, the current page is reloaded ✅

hellofromtonya replied:
Test report is here https://core.trac.wordpress.org/ticket/61330#comment:34

hellofromtonya replied:
Wait hold on -> Doesn't work in the modal 🔴 🐞 

Works in the plugin's card, but not in its "More Details" modal. cc @costdev

costdev committed:
Encode the referrer to preserve multiple query arguments.

Without URL-encoding, only the first `?arg=value` is retained, and any `&argX=valueX` is lost. Wrapping `wp_get_referer()` in `urlencode()` preserves these additional query arguments.
